recipe/2024-06-private-bicep-registries.md

```diff
@@ -0,0 +1,266 @@
+
+
+
+### Design details
+Currently, OCI-compliant registries are used to store Bicep recipes, with the ORAS client facilitating operations like publish and pull these recipes from the registries. ORAS provides package auth, enabling secure client authentication to remote registries. Private Registry credentials information i.e username and password must be stored in an `Application.Core/secretStores` resource and the secret resource ID is added to the recipe configuration.
+
+During the recipe deployment bicep driver checks for secrets information for that container registry the recipe deploying is stored on,  and creates an oras auth client to authenticate the private bicep registries.
+
+#### Different container registry providers provide different authenticate methods to login.
+
+***Azure***:
+
+Azure provides multiple ways to authenticate ACR using ORAS, we can use different types of credentials depending on the setup and preferences. Here are the main types of credentials to use:
+
+#### Admin User Credentials:
+Users can enable the admin user and obtain these credentials from the Azure portal:
+```
+Username : The ACR admin username.
+Password : The ACR admin password.
+```
+
+#### Service Principal:
+```
+Username: The service principal's client ID.
+Password: The service principal's client secret.
+```
+
+#### Azure CLI Token:
+```
+Username: Use '00000000-0000-0000-0000-000000000000' (a placeholder indicating token-based authentication).
+Password: Use the token obtained from az acr login.
+```
+
+You can obtain the token using the Azure CLI:
+```
+az acr login --name <registry-name> --expose-token
+```
+***Github***:
+Github uses personal access token as password to auth into ghcr.
+```
+Username: Github username.
+Password: Personal access token.
+```
+
+***AWS***:
+```
+Username: AWS
+Password: output of `aws ecr get-login-password --region region`
+```
+While there are various container registry providers, each offering multiple authentication methods, by leveraging ORAS, we abstract away the need for provider-specific code to handle authentication. 
+
+ORAS auth client example to authenticate private registry using username and password,
+```
+	repo.Client = &auth.Client{
+		Client: retry.DefaultClient,
+		Credential: auth.StaticCredential("orasregistry.azurecr.io", auth.Credential{
+			Username: "<username>",
+			Password: "<password>",
+		}),
+	}
+	repo.Client = client
+```
+### API design (if applicable)
+***Model changes***
+
+environment.tsp
+```diff
+@doc("Configuration for Recipes. Defines how each type of Recipe should be configured and run.")
+model RecipeConfigProperties {
+  @doc("Configuration for Terraform Recipes. Controls how Terraform plans and applies templates as part of Recipe deployment.")
+  terraform?: TerraformConfigProperties;
+
++  @doc(Configuration for Bicep Recipes. Controls how Bicep plans and applies templates as part of Recipe deployment.)
++  bicep?: BicepConfigProperties;
```

Comment:
Should we have a different name like recipeConfig or something like that? **Bicep** may be too general.

Reply:
as discussed , the parent property is already named as recipeConfig, so naming the child only as `Bicep`